sn_transfers/src/transfers/unsigned_transaction.rs

```diff
@@ -0,0 +1,308 @@
+pub struct UnsignedTransaction {
+    /// Output CashNotes stripped of their parent spends, unuseable as is
+    output_cashnotes_without_spends: Vec<CashNote>,
+    /// Change CashNote stripped of its parent spends, unuseable as is
+    change_cashnote_without_spends: Option<CashNote>,
+    /// Spends waiting to be signed along with their secret derivation index
+    spends: Vec<(Spend, DerivationIndex)>,
+}
+
+impl Debug for UnsignedTransaction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UnsignedTransaction")
+            .field(
+                "spends",
+                &self.spends.iter().map(|(s, _)| s).collect::<Vec<_>>(),
+            )
+            .finish()
+    }
+}
+
+impl UnsignedTransaction {
+    /// Create a new `UnsignedTransaction` with the given inputs and outputs
+    pub fn new(
+        available_cash_notes: Vec<CashNote>,
+        recipients: Vec<(NanoTokens, MainPubkey, DerivationIndex, OutputPurpose)>,
+        change_to: MainPubkey,
+        input_reason_hash: SpendReason,
+    ) -> Result<Self> {
+        // check total output amount
+        let total_output_amount = recipients
+            .iter()
+            .try_fold(NanoTokens::zero(), |total, (amount, _, _, _)| {
+                total.checked_add(*amount)
+            })
+            .ok_or(TransferError::ExcessiveNanoValue)?;
+
+        // pick input cash notes
+        let (chosen_input_cn, _theory_change_amount) =
+            select_inputs(available_cash_notes, total_output_amount)?;
+
+        // create empty output cash notes for recipients
+        let outputs: Vec<(CashNote, NanoTokens, OutputPurpose)> = recipients
+            .iter()
+            .map(|(amount, main_pk, derivation_index, purpose)| {
+                let cn = CashNote {
+                    parent_spends: BTreeSet::new(),
+                    main_pubkey: *main_pk,
+                    derivation_index: *derivation_index,
+                };
+                (cn, *amount, purpose.clone())
+            })
+            .collect();
+
+        // distribute value from inputs to output cash notes
+        let mut spends = Vec::new();
+        let mut change_cn = None;
+        let mut outputs_iter = outputs.iter();
+        for input in chosen_input_cn {
+            let input_key = input.unique_pubkey();
+            let input_value = input.value();
+            let input_ancestors = input
+                .parent_spends
+                .iter()
+                .map(|s| *s.unique_pubkey())
+                .collect();
+            let mut input_remaining_value = input_value.as_nano();
+            let mut donate_to = BTreeMap::new();
+
+            // take value from input and distribute it to outputs
+            while input_remaining_value > 0 {
+                if let Some((output, amount, purpose)) = outputs_iter.next() {
+                    let amount_to_take = min(input_remaining_value, amount.as_nano());
```

Comment:
if the current `output.amount > input_remaining_value`, then here we got `input_remaining_value` sent to that output recipient.
BUT, with the next input, the `outputs_iter.next()` will jump to the next recipient.
Which acutally the previous recipient is short of `output.amount - input_remaining_value` to be paid.

Reply:
Excellent find! Now fixed in the last commit!